#82 - edit profile features

kamilpiech97 opened the pull request.

kamilpiech97 committed:
Profile controller

kamilpiech97 committed:
Profile controller

kamilpiech97 committed:
fix birthday field and add some test for profile controller

kamilpiech97 committed:
fix birthday field and add some test for profile controller

kamilpiech97 committed:
password changing with old password confirm feature

kamilpiech97 committed:
password changing with old password confirm feature

kamilpiech97 committed:
email change feature, tests for this feature - to refractor and psalm check in next commmit

kamilpiech97 committed:
email change feature, tests for this feature - to refractor and psalm check in next commmit

kamilpiech97 committed:
Update services.php

hide API credentials

kamilpiech97 committed:
Update services.php

hide API credentials

kamilpiech97 committed:
avatar update, tests and some fixes in code

kamilpiech97 committed:
avatar update, tests and some fixes in code

krzysztofrewak committed:
#58 - updated laravel/(socialite, nova, framework), ecs, sass (#210)

kamilpiech97 committed:
fix classes, code ans style

kamilpiech97 committed:
fix classes, code ans style

krzysztofrewak commented:
Checks are not passing. Look at this: https://github.com/brewmap/brewmap/pull/211/checks?check_run_id=1698574175 

Without it we are not going to start any code review.

kamilpiech97 commented:
okey, I will fix it.

dependabot-preview[bot] committed:
#58 - (php) [Security] Bump laravel/framework from 8.22.0 to 8.22.1 (#212)

Bumps [laravel/framework](https://github.com/laravel/framework) from 8.22.0 to 8.22.1. **This update includes a security fix.**
- [Release notes](https://github.com/laravel/framework/releases)
- [Changelog](https://github.com/laravel/framework/blob/8.x/CHANGELOG-8.x.md)
- [Commits](https://github.com/laravel/framework/compare/v8.22.0...v8.22.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

kamilpiech97 committed:
rebase and fix composer.lock

kamilpiech97 committed:
rebase and fix composer.lock

krzysztofrewak commented on .vscode/settings.json:
```diff
@@ -0,0 +1,2 @@
+{
```
Please add `.vscode` directory to `.gitignore`.

krzysztofrewak commented on .vscode/settings.json:
```diff
@@ -0,0 +1,2 @@
+{
+}
```
Alwats take care about empty lines at the end of the file. You're probably working on Windows, right?

kamilpiech97 replied:
yep, Windows ;p

krzysztofrewak commented on app/Eloquent/Profile.php:
```diff
@@ -20,8 +21,23 @@ class Profile extends Model
     }
+
+    /**
+     * @psalm-suppress MissingThrowsDocblock
```
You missed the point of my comment in #82. `@psalm-suppress` is supressing Psalm inspections for docblocked class/method/LOC. If you would like to supress it in all your classes, you should add global suppression in Psalm configuration file. But documenting throwables is something we would like to keep.

So the only solution is to add docblock in question (do the same for all other suppressions):
```
    /**
     * @throws InvalidFormatException
     */
    public function setBirthdayAttribute($value): void
    {
        $this->attributes["birthday"] = Carbon::parse($value);
    }
```

kamilpiech97 replied:
```php
@throws `InvalidFormatException
```
When I add above annonation to **Profile** class, 
Psalm after check code give me error like this:

```php
ERROR: UndefinedDocblockClass - app/Eloquent/Profile.php:32:16 - Docblock-defined class or interface Brewmap\Eloquent\InvalidFormatException does not exist (see https://psalm.dev/200)
  @throws InvalidFormatException
  ```

krzysztofrewak replied:
Look at the namespace again.

krzysztofrewak commented on app/Eloquent/Profile.php:
```diff
@@ -20,8 +21,23 @@ class Profile extends Model
+
+    public function getAvatarPathAttribute($value): string
+    {
+        return Storage::url($value);
```
We are trying to avoid facades. Also: always type your parameters.

krzysztofrewak commented on app/Exceptions/User/NewEmailChangingException.php:
```diff
@@ -0,0 +1,14 @@
+
+class NewEmailChangingException extends ApiException
+{
+    protected $message = "This email is being used by another user!";
```
Exclamations are overkill. ;)

krzysztofrewak replied:
Another question is: do we really would like to inform users about some other users emails in our system?

kamilpiech97 replied:
fixed and removed exclamations ;))

krzysztofrewak commented on app/Http/Controllers/API/Profile/ProfileController.php:
```diff
@@ -0,0 +1,36 @@
+
+class ProfileController extends Controller
+{
+    /**
```
We don't need these comments.

krzysztofrewak commented on app/Http/Controllers/API/Profile/ProfileController.php:
```diff
@@ -0,0 +1,36 @@
+    public function edit(): JsonResponse
+    {
+        return response()->json([
+            new ProfileResource(auth()->user()->profile),
```
Please inject user into your controller instead of using `auth()` helper.

krzysztofrewak commented on app/Http/Controllers/API/Profile/UploadAvatarController.php:
```diff
@@ -0,0 +1,27 @@
+
+class UploadAvatarController extends Controller
+{
+    private $pathToAvatar;
```
Always type your properties. Also: why it's private and not protected?

HerringTheCoder replied:
I think it might be a local variable even, because StoreFileService, which is currently the only provider of avatar path value, is only injected in the 'store' method scope.

Also, as a future bonus (inb4 yagni), exposing an interface instead of concrete StoreFileService sounds like a cool thing to do here, because current implementation is bound to underlying local storage. It could make integration testing easier too.

krzysztofrewak replied:
@kamilpiech97, why didn't you address @HerringTheCoder comments?

krzysztofrewak commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,42 @@
+    public function change(UpdateEmailRequest $request, UpdateEmailService $service): JsonResponse
+    {
+        $service->sendNotifyForNewEmail($request->email);
+        return response()->json([
```
To be honest (and this in not a strict rule) I would prefer separating return with empty line. But this is something to talk about.

krzysztofrewak commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,42 @@
+     * @param UpdateEmailRequest $request
+     * @psalm-suppress MissingThrowsDocblock
+     */
+    public function verify(Request $request, User $user, string $email, UpdateEmailService $service): JsonResponse
```
I don't understand why `$request` is type of `Illuminate\Http\Request`, but in docblock you're suggesting that's a `Brewmap\Http\Requests\User\UpdateEmailRequest` instance.

kamilpiech97 replied:
my fault, i changed it after ur comment.

krzysztofrewak commented on app/Http/Resources/Profile.php:
```diff
@@ -0,0 +1,26 @@
+class Profile extends JsonResource
+{
+    /**
+     * Transform the resource into an array.
```
Remove unneccessary docblock.

krzysztofrewak commented on app/Http/Resources/Profile.php:
```diff
@@ -0,0 +1,26 @@
+     */
+    public function toArray($request): array
+    {
+        return [
```
Add docblock to endure what we are doing here. Resources are dark Laravel magic and it's little confusing.

```
        /** @var Profile $profile */
        $profile = $this->resource;
```

And then: `"public_name" => $profile->public_name,`

krzysztofrewak commented on app/Http/Resources/Profile.php:
```diff
@@ -0,0 +1,26 @@
+        return [
+            "data" => [
+                "public_name" => $this->public_name,
+                "avatar_path" => $this->avatar_path,
```
Look at this: https://github.com/brewmap/brewmap/blob/develop/app/Eloquent/Traits/CamelCaseAttributes.php

We are remapping all properties into camelCased, so you should use `$this->avatarPath`.

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+use Illuminate\Support\Facades\URL;
+
+/**
+ * @psalm-suppress PropertyNotSetInConstructor
```
Probably it will be common problem for all notifications. Add this suppression in `psalm.xml`.

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+    use Queueable;
+
+    /**
+     * The user Email
```
Remove these comments.

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+     * @param mixed $notifiable
+     * @return array
+     */
+    public function via($notifiable)
```
Typing.

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    public function via($notifiable)
+    {
+        return ["mail"];
```
I can imagine the situation where we would have like 100 notification classes. I would really do not like to change string `"mail"` in random number of them.

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return string
+     */
+    protected function verifyRoute(AnonymousNotifiable $notifiable)
```
Typing!

krzysztofrewak commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    protected function verifyRoute(AnonymousNotifiable $notifiable)
+    {
+        return URL::temporarySignedRoute("api.email.change", now()->addMinutes(90), [
```
Move this magic number into some const.

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function sendNotifyForNewEmail(string $email): void
+    {
+        Notification::route("mail", $email)
```
Try to remove the facade.

mlencki replied:
Maybe `Illuminate\Notifications\ChannelManager` or `Illuminate\Notifications\Channels\MailChannel` will help

kamilpiech97 replied:
Works with `Illuminate\Notifications\AnonymousNotifiable` :)
`

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,48 @@
+    public function sendNotifyForNewEmail(string $email): void
+    {
+        Notification::route("mail", $email)
+            ->notify(new EmailChangeNotification(auth()->user()->id));
```
And think about using `auth()` here. What would happen with this service ran from queued event?

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function checkEmailBeforeUpdate(String $email): void
+    {
+        if (User::where("email", $email)->first() !== null) {
```
Always use `query()`static proxy for Eloquent queries. Eg. `User::query()->where(...)`; it's better for IDE.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,30 @@
+
+class MailSending implements Context
+{
+    protected $signedUrl;
```
Type it.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,30 @@
+     */
+    public function anEmailShouldBeSent(): void
+    {
+        Notification::assertSentTo(new AnonymousNotifiable(), EmailChangeNotification::class, function ($notification, $channels, $notifiable) {
```
Type anonymous functions too.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,30 @@
+
+    public function __construct()
+    {
+        Notification::fake();
```
I don't like this solution. I would prefer mocking it by standard mock service, but we can wait for someone else thoughts and comments.

mlencki replied:
Facade is using `Illuminate\Support\Testing\Fakes\NotificationFake`, maybe we can use it here directly?

kamilpiech97 replied:
I am trying use `Illuminate\Support\Testing\Fakes\NotificationFake` but when I checked assertion, function threw error...

krzysztofrewak replied:
Could you be more pricise? What kind of error? Maybe you could paste error message here?

krzysztofrewak commented on features/API/email.feature:
```diff
@@ -0,0 +1,35 @@
+  Scenario: User is attempting to update email address with busy email
+    Given user is logged in as "user@example.com"
+    And a user is requesting "api/email" using "PATCH"
+    And request body contains "email" equal "user@example.com.pl"
```
Use `@example.com` domain as it's restricted one.

krzysztofrewak commented on features/API/email.feature:
```diff
@@ -0,0 +1,35 @@
+    When a request is sent
+    Then a response status code should be "422"
+    And response body should contain:
+      | email         | The email has already been taken  |
```
There is something wrong with whitespaces here. Aren't there VSC plugin for Gherkin formatting?

krzysztofrewak commented on features/API/profile.feature:
```diff
@@ -0,0 +1,58 @@
+    And request body contains "file" equal "it_is_not_a_file"
+    When a request is sent
+    Then a response status code should be "422"
+
```
What happened here with empty lines? There should be exact one empty line at the EOF.

krzysztofrewak commented on features/API/profile.feature:
```diff
@@ -0,0 +1,58 @@
+    When a request is sent
+    Then a response status code should be "200"
+    And response body should contain:
+      | data         | {"public_name":"John Doe","avatar_path":"","birthday":"2000-01-01"}   |
```
This is really bad approach. Unpack this json and create test like:
```
    And response body "data" should contain:
      | public_name | John Doe   |
      | avatar_path |            | 
      | birthday    | 2000-01-01 |
```

krzysztofrewak commented on routes/api.php:
```diff
@@ -13,3 +17,15 @@
+
+$router->group([
+    "middleware" => "auth:sanctum",
+], function () use ($router): void {
```
Type router.

kamilpiech97 replied:
I have problem with that..
Can I type it?

HerringTheCoder replied:
Is Illuminate\Routing\Router type not sufficient?
It's already working with this line above:
`$router = app(Router::class);`

krzysztofrewak reviewed (changes requested):
Okay. You need to fix or adjust some things described in comments. Please remember that I didn't comment the same thing for every class.

Also: please add "Should close #82" sentence to your PR message. It's for GitHub Actions automation. You should also provide information what you have done in this PR.

mlencki commented on app/Services/StoreFileService.php:
```diff
@@ -0,0 +1,23 @@
+
+class StoreFileService
+{
+    protected Object $file;
```
Probably we can use `Illuminate\Http\UploadedFile` type here to be more specific
```suggestion
    protected UploadedFile $file;
```

mlencki commented on app/Services/StoreFileService.php:
```diff
@@ -0,0 +1,23 @@
+
+    public function storeFile(string $uploadFolder)
+    {
+        $this->path = $this->file->store($uploadFolder, "public");
```
You can try to inject `Illuminate\Contracts\Filesystem\Filesystem` abstract to the service.

Then we could simply switch through different storage drivers just by simply changing .env variable `FILESYSTEM_DRIVER` to `public`, `local`, `s3` etc.

You can see it in `filesystems.php` config file
https://github.com/brewmap/brewmap/blob/021d7de73f09bcecf104f397a60b151820ca5600/config/filesystems.php#L6
Laravel will use it to provide proper `Filesystem` implementation for selected storage driver.

mlencki commented on app/Services/StoreFileService.php:
```diff
@@ -0,0 +1,23 @@
+{
+    protected Object $file;
+
+    public function __construct(Request $request)
```
I would rather not pass entire `Request` object to the service as we only need the uploaded file here.

You can also try to inject `Illuminate\Contracts\Filesystem\Filesystem` as I mentioned in another comment.

kamilpiech97 replied:
I am inject directly file into function, like this:
```
public function storeFile(string $uploadFolder, UploadedFile $file): string
    {
        $this->path = $file->store($uploadFolder);

        return $this->path;
    }
```

krzysztofrewak replied:
I have no idea what comment above means.

kamilpiech97 replied:
Sorry,  I will explain more.
I have one method and constructor now in **StoreFileService.php**.
I will put only one method (**storeFile**) to **StoreFIleService** because I inject file to that method from controller method. 
Code below.

UploadAvatarController:
```
public function store(UpdateProfileAvatarRequest $request, StoreFileService $service): JsonResponse
    {
        $this->pathToAvatar = $service->storeFile("users", $request->file("file"));

        return new JsonResponse([
            "path_to_file" => $this->pathToAvatar,
        ]);
    }
```

StoreFileService:
```
public function storeFile(string $uploadFolder, UploadedFile $file): string
    {
        $this->path = $file->store($uploadFolder);

        return $this->path;
    }
```

mlencki commented on app/Services/UpdatePasswordService.php:
```diff
@@ -0,0 +1,31 @@
+        if (!$this->hash->check($input["old_password"], auth()->user()->password)) {
+            throw new OldPasswordMismatchException();
+        }
+        auth()->user()->update([
+            "password" => $this->hash->make($input["password"]),
+        ]);
```
You can try to inject `Illuminate\Contracts\Auth\Guard` the same way like you're doing with `Illuminate\Contracts\Hashing\Hasher`.

Then you could try to call `user()` on `guard` object instead using `auth()` helper function
```suggestion
        $this->guard->user()->update([
            "password" => $this->hash->make($input["password"]),
        ]);
```

mlencki commented on app/Services/UpdateProfileService.php:
```diff
@@ -0,0 +1,13 @@
+{
+    public function updateProfileData(array $input): void
+    {
+        auth()->user()->profile->update($input);
```
Here, you can also try to use `Illuminate\Contracts\Auth\Guard`
```suggestion
        $this->guard->user()->profile->update($input);
```

mlencki commented on app/Notifications/EmailChangeNotification.php:
```diff
@@ -0,0 +1,71 @@
+        return ["mail"];
+    }
+
+    /**
+     * Get the mail representation of the notification.
+     *
+     * @param mixed $notifiable
+     * @return \Illuminate\Notifications\Messages\MailMessage
+     */
+    public function toMail($notifiable)
```
```suggestion
    public function toMail($notifiable): MailMessage
```

mlencki commented on app/Http/Controllers/API/User/PasswordChangeController.php:
```diff
@@ -0,0 +1,27 @@
+    public function update(UpdatePasswordRequest $request, UpdatePasswordService $service): JsonResponse
+    {
+        $service->updatePassword($request->validated());
+        return response()->json([
```
You can try to use `Iluminate\Contracts\Routing\ResponseFactory` or just simply return new `JsonResponse` instance instead of using `response()` helper.

```suggestion
        return $this->responseFactory->json([
```

```suggestion
        return new JsonResponse([
```

kamilpiech97 replied:
It will be change in all controllers for API?

krzysztofrewak replied:
Of course. We are doing everything one way.

mlencki commented on app/Exceptions/User/OldPasswordMismatchException.php:
```diff
@@ -0,0 +1,14 @@
+
+class OldPasswordMismatchException extends ApiException
+{
+    protected $message = "Old password is mismatch!";
```
There's no need to yell at our users :wink: 

Also `Old password is mismatched.` or `Old password doesn't match.` wouldn't be more grammatically correct?
```suggestion
    protected $message = "Old password doesn't match.";
```

kamilpiech97 committed:
changes in code after code review

kamilpiech97 committed:
changes in code after code review

kamilpiech97 committed:
ecs fix

kamilpiech97 committed:
ecs fix

kamilpiech97 committed:
fix code

kamilpiech97 committed:
fix code

krzysztofrewak commented on .gitignore:
```diff
@@ -16,3 +16,5 @@ yarn-error.log
 /.psalm
 /.npm
+
+.vscode
```
And again there's no empty line at EOF.

krzysztofrewak commented on app/Eloquent/Profile.php:
```diff
@@ -20,8 +21,26 @@ class Profile extends Model
+    /**
+     * @psalm-suppress MissingThrowsDocblock
+     */
+    public function getAvatarPathAttribute(string $value): string
```
Add docblocks properties above the class for all magic Laravalesque things.

kamilpiech97 replied:
I resolved this but not commit my changes yet.

krzysztofrewak replied:
Please remember to import all classes.

krzysztofrewak commented on app/Exceptions/User/NewEmailChangingException.php:
```diff
@@ -0,0 +1,14 @@
+
+class NewEmailChangingException extends ApiException
+{
+    protected $message = "Please choose another email address.";
```
You should not put instructions like it here. Message is a place for excpetion message only.

kamilpiech97 replied:
Okay, i will delete message and return only http code?

krzysztofrewak replied:
No, you're getting me wrong. The message should be somehting like "You cannot choose privoded e-mail address".

kamilpiech97 replied:
Aaa, okey. Now I know what u want 😅

krzysztofrewak commented on app/Http/Controllers/API/Profile/ProfileController.php:
```diff
@@ -0,0 +1,32 @@
+            new ProfileResource($request->user()->profile)
+        );
+    }
+
```
There's a lot of blank line here.

krzysztofrewak commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * @throws
```
Remove unnecesary comments.

krzysztofrewak commented on app/Http/Resources/Profile.php:
```diff
@@ -0,0 +1,27 @@
+class Profile extends JsonResource
+{
+    /**
+     * @param \Illuminate\Http\Request $request
```
Please import all your classes.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @return string
+     */
+    protected function verifyRoute(AnonymousNotifiable $notifiable)
```
Remove docblock and provide return type.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+{
+    protected string $userId;
+    protected string $verifyUrl;
+    protected Object $time;
```
`Object`? What are we?

kamilpiech97 replied:
Now, Carbon.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+     */
+    protected function verifyRoute(AnonymousNotifiable $notifiable)
+    {
+        $this->verifyUrl = URL::temporarySignedRoute("api.email.change", $this->time, [
```
Why are you assiging this url into class' field? And why are you using facade again?

kamilpiech97 replied:
`Illuminate\Contracts\Routing\UrlGenerator` works very good, and sorry for facade.
Facades are addictive... ;c

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+    protected function verifyRoute(AnonymousNotifiable $notifiable)
+    {
+        $this->verifyUrl = URL::temporarySignedRoute("api.email.change", $this->time, [
+            "user" => $this->userId,
```
I don't like idea of passing something called `$user` and being... string?

kamilpiech97 replied:
I pass here user id if I pass old email it will be okay? Id and email is uniq in user model.
I think it will work good.

kamilpiech97 replied:
Or name this variable like user_id or smth.

krzysztofrewak replied:
Changing varaiable name would be enough I think.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @param mixed $notifiable
+     */
+    public function toMail($notifiable): MailMessage
```
Again: remove docblock and provide proper type hint. Also: why `mixed`?

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,50 @@
+    public function __construct(string $userId)
+    {
+        $this->userId = $userId;
+        $this->time = config("constants.notification.time_to_change_email");
```
I think that word "constant" is quite far away of "current time + 90 minutes". You cannot put values like that into configuration - it will be cached on production environment and will fail fo this task.

kamilpiech97 replied:
Instead of const, passing `now()->addMinutes(90)` or `90` number will be okay?

krzysztofrewak replied:
Something like this? 
```suggestion
        $this->time = Carbon::now()->addMinutes(config("notifications.email_change_timeout"));
```

kamilpiech97 replied:
Yep, it looks great! 🙂

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * @psalm-suppress MissingThrowsDocblock
+     */
+    public function updateEmail(User $user, String $email): void
```
In PHP it's a lowercase `string`.

krzysztofrewak commented on config/constants.php:
```diff
@@ -0,0 +1,9 @@
+
+return [
+    "notification" => [
+        "time_to_change_email" => now()->addMinutes(90),
```
Probably you could store here `90` as number only.

krzysztofrewak reviewed (changes requested):
Okay, there's a lot to improve. I resolved some comments from previous review, but some important ones are still opened. I also add a few new comments to your recent changes.

Next time, please finish all fixes or ask for help. There are many things that we pointed out last time (like facades, typing, suppressing Psalm magic) that you're repeating.

kamilpiech97 committed:
fix code after PR, typing etc.

kamilpiech97 committed:
fix code after PR, typing etc.

krzysztofrewak commented on app/Exceptions/User/NewEmailChangingException.php:
```diff
@@ -0,0 +1,14 @@
+
+class NewEmailChangingException extends ApiException
+{
+    protected $message = "You cannot choose privoded e-mail address.";
```
Typo.

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,50 @@
+{
+    protected AnonymousNotifiable $notify;
+
+    public function __construct(AnonymousNotifiable $notify)
```
I still don't get it. What the heck is `AnonymousNotifiable`? Why its instance is called `$notify`? You cannot name your variables with verbs, because they are not doing anything.

kamilpiech97 replied:
**anonymousNotifiable** or **notification** will be okay ?

mlencki replied:
Since ` User ` model is already using `Notifable` trait
https://github.com/brewmap/brewmap/blob/021d7de73f09bcecf104f397a60b151820ca5600/app/Eloquent/User.php#L41

I would prefer to use notifications like it is explained in Laravel documentation:
https://laravel.com/docs/8.x/notifications#using-the-notifiable-trait

So you could send notifications to given user just by simply:
```php
$user->notify(new EmailChangeNotification());
```
Then you won't have to use  `AnonymousNotifiable` and things like `user_id` or `email` will be accessible in `$notifiable` e.g.
```php
protected function verifyRoute(User $notifiable): string
{
       return app()->make(UrlGenerator::class)
            ->temporarySignedRoute("api.email.change", $this->time, [
                "user_id" => $user->id,
                "email" => $user->email
       ]);
}
```

kamilpiech97 replied:
but I send notification to new email not already used email by User.

krzysztofrewak replied:
I believe that you could do something like this:
```php
$user->email = $newEmail;
$user->notify(new EmailChangeNotification());
```

kamilpiech97 replied:
I must send notification with link to update new email and i must have user id or old email.

kamilpiech97 replied:
temporarySignedRoute create signature based on new email and user id.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,28 @@
+     */
+    public function anEmailShouldBeSent(): void
+    {
+        Notification::assertSentTo(new AnonymousNotifiable(), EmailChangeNotification::class, function (EmailChangeNotification $notification, array $channels, AnonymousNotifiable $notifiable) {
```
You can use arrow function for that:
```suggestion
       Notification::assertSentTo(new AnonymousNotifiable(), EmailChangeNotification::class, fn (EmailChangeNotification $notification, array $channels, AnonymousNotifiable $notifiable): bool => $notifiable->routes["mail"] === "new_email@example.com");
```

You forgot about return type. You didn't remove facade too.

kamilpiech97 replied:
In this case with facade, I am waiting for help from @mlencki.

mlencki replied:
It was failing with `550 relaying denied` message, probably because it was still trying to use real notification dispatcher in tests.

Nothing was mocked when you just simply instantiated `NotificationFake` and tried to perform assertions.

You need to swap manually default implementation of  `Illuminate\Contracts\Notifications\Dispatcher` somewhere in tests.

You can do it in context's constructor or some additional step e.g.
```php
$this->notificatonFake = new NotificationFake();

app()->bind(Dispatcher::class, function ($app) {
       return $this->notificatonFake;
});

```

so then you could use it in steps e.g.
```php
/**
* @Then an email should be sent
*/
public function anEmailShouldBeSent(): void
{
      $this->notificatonFake->assertSentTo(.....)
}
```

or you could create new `TestServiceProvider` where you could bind `NotificationFake` the same way

kamilpiech97 replied:
okey, i will try with it.

kamilpiech97 replied:
it works, thank you so much.

krzysztofrewak replied:
`TestingServiceProvider` with binding mocks is not a good option. Soon it will be huge and it will mock a milion classes for every test.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,48 @@
+        $this->userId = $userId;
+        $this->time = Carbon::now()->addMinutes(config("notifications.email_change_timeout"));
+    }
+
```
Please check all your files for these double blank lines.

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @throws NewEmailChangingException
+     */
+    public function checkEmailBeforeUpdate(String $email): void
```
One more `String`? Oh, come on, you have already changed it on method above.

krzysztofrewak reviewed (changes requested):
Please recheck my previous code reviews. There're five more comments to fix and resolve.

mlencki commented on app/Http/Controllers/API/Profile/ProfileController.php:
```diff
@@ -0,0 +1,30 @@
+
+class ProfileController extends Controller
+{
+    public function edit(Request $request): JsonResponse
```
Shouldn't it be named `get`? It doesn't edits anything but only retrieves authenticated user's profile.

kamilpiech97 replied:
i think, edit function get data do edit or smth, i can change that.

mlencki commented on routes/api.php:
```diff
@@ -13,3 +17,15 @@
+$router->group([
+    "middleware" => ["auth:sanctum"],
+], function (Router $router): void {
+    $router->get("/profile", [ProfileController::class, "edit"]);
```
Please rename controller's method to `get` or something as it doesn't edits anything.

mlencki commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,36 @@
+{
+    public function change(UpdateEmailRequest $request, UpdateEmailService $service): JsonResponse
+    {
+        $service->sendNotifyForNewEmail($request->email, $request->user()->id);
```
`send` or `notify`? I think method name with single verb will be better e.g. **notify**AboutNewEmail or **send**NotificationAboutNewEmail`

mlencki commented on resources/lang/en/profile.php:
```diff
@@ -0,0 +1,10 @@
+return [
+    "profile_updated" => "Successfully updated profile data!",
+    "password_updated" => "Successfully updated password!",
+    "notification_new_email" => "Notification successfully sended for your new email! Please confirm on new email your email address.",
```
Isn't `sent` more accurate past tense of `send` than `sended`? Also check other files, please
```suggestion
    "notification_new_email" => "Notification about your new e-mail has been sent! Please confirm your new e-mail address.",
```

mlencki commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,50 @@
+     */
+    public function checkEmailBeforeUpdate(String $email): void
+    {
+        if (User::query()->where("email", $email)->first() !== null) {
```
What about using `exists()` here?
```suggestion
        if (User::query()->where("email", $email)->exists()) {
```

kamilpiech97 committed:
changed notificationFake for tests(without facades), added StoreFile interface, fixed double lines mistakes in code, added not-imported classes and some typo or grammar mistakes

kamilpiech97 committed:
changed notificationFake for tests(without facades), added StoreFile interface, fixed double lines mistakes in code, added not-imported classes and some typo or grammar mistakes

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,50 @@
+        $this->anonymousNotifiable = $anonymousNotifiable;
+    }
+
+    public function notifyAboutNewEmail(string $email, string $id): void
```
Why `notifyAboutNewEmail()` and `updateEmail()` methods are taking the same arguments but in different order? And why yo're expecting string id? Maybe `User` instance would be better?

kamilpiech97 replied:
I changed that like this. I pass **User** to the service function and I am getting id form User instance. It looks beeter.
```
public function notifyAboutNewEmail(User $user, string $email): void
    {
        $this->anonymousNotifiable
            ->route("mail", $email)
            ->notify(new EmailChangeNotification($user->id));
    }
```

krzysztofrewak commented on app/Services/UpdateEmailService.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @throws NewEmailChangingException
+     */
+    public function updateEmail(string $user_id, string $email): void
```
User instance would be better here for sure.

krzysztofrewak commented on app/Contracts/StoreFile.php:
```diff
@@ -0,0 +1,12 @@
+
+interface StoreFile
+{
+    public function storeFile(string $uploadFolder, UploadedFile $file);
```
Why there's no return type here?

krzysztofrewak commented on app/Eloquent/Profile.php:
```diff
@@ -20,8 +24,27 @@ class Profile extends Model
     protected $primaryKey = "user_id";
 
+    protected $fillable = ["public_name", "avatar_path", "birthday"];
+
```
This empty line is unnecessary. Look at this: https://www.php-fig.org/psr/psr-12/#43-properties-and-constants

krzysztofrewak commented on app/Http/Controllers/API/Profile/UploadAvatarController.php:
```diff
@@ -0,0 +1,27 @@
+use Illuminate\Http\JsonResponse;
+
+/**
+ * @psalm-suppress MissingConstructor
```
Remove this suppression, it's useless in this case.

krzysztofrewak commented on app/Http/Controllers/API/Profile/UploadAvatarController.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class UploadAvatarController extends Controller
+{
+    protected string $pathToAvatar;
```
Remove this class' property and move it to method's body.

krzysztofrewak commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,36 @@
+{
+    public function change(UpdateEmailRequest $request, UpdateEmailService $service): JsonResponse
+    {
+        $service->notifyAboutNewEmail($request->email, $request->user()->id);
```
Please use `$request->get("email")` insttead of `$request->email`.

krzysztofrewak commented on app/Http/Controllers/API/User/EmailChangeController.php:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * @throws NewEmailChangingException
+     */
+    public function verify(Request $request, string $user_id, string $email, UpdateEmailService $service): JsonResponse
```
Use camel case.

krzysztofrewak commented on app/Notifications/Mail/EmailChangeNotification.php:
```diff
@@ -0,0 +1,47 @@
+    public function __construct(string $userId)
+    {
+        $this->userId = $userId;
+        $this->time = Carbon::now()->addMinutes(config("notifications.email_change_timeout"));
```
Rename it to `$timeout` or something.

krzysztofrewak replied:
:eyes:

krzysztofrewak commented on app/Notifications/MailNotification.php:
```diff
@@ -0,0 +1,20 @@
+
+    public function via(AnonymousNotifiable $notifiable): array
+    {
+        return ["mail"];
```
You are not even using this `$notifiable` variable. I still don't get why it's `AnonymousNotifiable` used here.

krzysztofrewak commented on app/Services/StoreFileService.php:
```diff
@@ -0,0 +1,26 @@
+    protected string $path;
+
+    /**
+     * @return string
```
Remove this useless comment.

krzysztofrewak commented on app/Services/StoreFileService.php:
```diff
@@ -0,0 +1,26 @@
+use Illuminate\Http\UploadedFile;
+
+/**
+ * @psalm-suppress MissingConstructor
```
Remove useless suppression.

krzysztofrewak replied:
:eyes:

krzysztofrewak commented on app/Services/UpdatePasswordService.php:
```diff
@@ -0,0 +1,34 @@
+
+declare(strict_types=1);
+
+namespace Brewmap\Services;
```
Move these services to some new namespace like `Brewmap/Services/User` or something like that.

krzysztofrewak commented on app/Services/UpdatePasswordService.php:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * @throws OldPasswordMismatchException
+     */
+    public function updatePassword(array $input): void
```
For me, `updatePassword` should take User instance and new password only.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,35 @@
+    public function notificationFakeIsMocked(): void
+    {
+        $this->notificatonFake = new NotificationFake();
+        app()->bind(Dispatcher::class, fn (Application $app): NotificationFake => $this->notificatonFake);
```
You are not even using this `Application $app` here.

krzysztofrewak commented on app/Testing/Context/MailSending.php:
```diff
@@ -0,0 +1,35 @@
+    public function anEmailShouldBeSent(): void
+    {
+        $this->notificatonFake
+            ->assertSentTo(new AnonymousNotifiable(), EmailChangeNotification::class, fn (EmailChangeNotification $notification, array $channels, AnonymousNotifiable $notifiable): bool => $notifiable->routes["mail"] === "new_email@example.com");
```
Move string to some constant.

krzysztofrewak commented on composer.json:
```diff
@@ -19,7 +19,7 @@
     "fideloper/proxy": "^4.2",
     "fruitcake/laravel-cors": "^2.0",
-    "guzzlehttp/guzzle": "^7.0.1",
+    "guzzlehttp/guzzle": "^7.0",
```
Why did you change that?

kamilpiech97 replied:
Sorry, I think when I was installing something for email testing and new depency downgrade **guzzlehttp**. I fixed that.

krzysztofrewak reviewed (changes requested):
Remember about ones above.

dependabot-preview[bot] committed:
#58 - (php) Bump laravel/framework from 8.22.1 to 8.25.0 (#224)

Bumps [laravel/framework](https://github.com/laravel/framework) from 8.22.1 to 8.25.0.
- [Release notes](https://github.com/laravel/framework/releases)
- [Changelog](https://github.com/laravel/framework/blob/8.x/CHANGELOG-8.x.md)
- [Commits](https://github.com/laravel/framework/compare/v8.22.1...v8.25.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

kamilpiech97 committed:
fix services, make binding for contract and add User model for new email verify

kamilpiech97 committed:
fix services, make binding for contract and add User model for new email verify

kamilpiech97 committed:
Merge branch 'develop' into #82-edit-profile-features

mlencki commented on app/Services/User/UpdatePasswordService.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * @throws OldPasswordMismatchException
+     */
+    public function index(string $oldPassword, string $newPassword, User $user): void
+    {
+        $this->checkOldPassword($oldPassword, $user);
+        $this->updatePassword($newPassword, $user);
+    }
+
+    /**
+     * @throws OldPasswordMismatchException
+     */
+    public function checkOldPassword(string $oldPassword, User $user): void
+    {
+        if (!$this->hash->check($oldPassword, $user->password)) {
+            throw new OldPasswordMismatchException();
+        }
+    }
+
+    public function updatePassword(string $newPassword, User $user): void
+    {
+        $user->update([
+            "password" => $this->hash->make($newPassword),
+        ]);
+    }
```
why `index`? This name can be confusing, please change to `updatePassword` or something
```suggestion
    public function updatePassword(string $oldPassword, string $newPassword, User $user): void
    {
        $this->checkOldPassword($oldPassword, $user);
        
        $user->update([
            "password" => $this->hash->make($newPassword),
        ]);
    }

    /**
     * @throws OldPasswordMismatchException
     */
    public function checkOldPassword(string $oldPassword, User $user): void
    {
        if (!$this->hash->check($oldPassword, $user->password)) {
            throw new OldPasswordMismatchException();
        }
    }
```

mlencki commented on app/Http/Controllers/API/Profile/UploadAvatarController.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public function store(UpdateProfileAvatarRequest $request, StoreFile $service): JsonResponse
+    {
+        /** @var string $pathToAvatar */
```
This is probably unnecessary as there is already return type in `StoreFile` interface:
https://github.com/brewmap/brewmap/blob/a683f75c7315f85f6ac951278f39adb4ffd08445/app/Contracts/StoreFile.php#L11
```suggestion
```

mlencki reviewed (commented):
Looks quite good 👍 Only 2 small things to fix. Also, please recheck unresolved @krzysztofrewak comments, especially the`AnonymousNotifiable` thing.

kamilpiech97 committed:
2 fixes for latest mlencki comments

kamilpiech97 committed:
2 fixes for latest mlencki comments

kamilpiech97 committed:
replace the time variable with timeout and delete useless suppression in StoreFileService

kamilpiech97 committed:
replace the time variable with timeout and delete useless suppression in StoreFileService

kamilpiech97 committed:
Merge branch '#82-edit-profile-features' of https://github.com/brewmap/brewmap into #82-edit-profile-features

kamilpiech97 committed:
changes after rebase

kamilpiech97 committed:
changes in dependabot

kamilpiech97 committed:
Merge branch 'develop' into #82-edit-profile-features
